aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -385,7 +386,11 @@ Tensor & index_copy_(Tensor & self, int64_t dim, const Tensor & index, const Ten
                    source.dim(), "), destination dimensionality (", self.dim(), ")");
   }
 
-  TORCH_CHECK_INDEX(index.scalar_type() == ScalarType::Long, "index_copy_(): Expected LongTensor for index");
+  TORCH_CHECK(index.scalar_type() == ScalarType::Long, "index_copy_(): Expected a long tensor for index, but got ", index.scalar_type())
+  TORCH_CHECK(self.scalar_type() == source.scalar_type(), "index_copy_(): self and source expected to have the same dtype, but got (self) ", self.scalar_type(), " and (source) ", source.scalar_type());
```

Comment:
I wonder if we should relax this condition to allow copying ints to floating point tensors and floats to complex tensors etc. 
```
>>> t = torch.tensor([[1, 2, 3], [4, 5, 6], [7, 8, 9]], dtype=torch.int)
>>> index = torch.tensor([0, 4, 2])
>>> x.index_copy_(0, index, t)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Expected object of scalar type Float but got scalar type Int for argument #4 'source' in call to _th_index_copy_
```
I propose relaxing the existing condition, because we do allow similar behavior in `torch.Tensor.copy_()`:
```
>>> x=torch.zeros(2,3, dtype=torch.int)
>>> y=torch.randn(2,3)
>>> y.copy_(x)
tensor([[0., 0., 0.],
        [0., 0., 0.]])
```
cc. @ngimel

Reply:
I think this makes sense. Relaxing this to `at::can_cast(source, self)` would also make the preconditions on these operations consistent with the `out=` rules. At the moment, the `out=` version of `torch.take` has this `self.scalar_type() == source.scalar_type()` kind of check, which is more restrictive than the `out=` operations in `torch.linalg`.

The only problem is that, for this to be consistent, it would have to be implemented across all the methods for advanced indexing, which will take some time.

Reply:
Maybe a good issue to follow-up with, @anjali411? I'd prefer this PR not add or change functionality since it's principally about porting the operation (although fixing bugs would be fine).

Reply:
@mruberry yeah that makes sense! @Lezcano could you file an issue for that?

Reply:
Done in https://github.com/pytorch/pytorch/issues/53257